### ISSUE TITLE
[BUG] Importing torii-relay into a new rust library leads to a compilation error

### ISSUE THREAD

**Issue by elton-cs** [dojoengine/dojo#2100]
**Describe the bug**
Importing `torii-relay` or any crate that depends on it (I was using `torii-client`) leads to a compilation error when running `cargo build`. Oddly enough, this error doesn't seem to show up when building the entire dojo library.

**To Reproduce**
- Create a new rust library: `cargo new --lib {LIBRARY_NAME}`
- Add torii-relay as dependency in Cargo.toml file: 
```
[dependencies]
torii-relay = {git = "https://github.com/dojoengine/dojo.git" }
```
- Build library: `cargo build`

**Expected behavior**
The library should build without error.

**Screenshots**
![image](https://github.com/dojoengine/dojo/assets/123023258/e3c8763c-b2e7-4445-aaab-27f20e742827)

**Additional context**
It looks like the libp2p dependency of this crate recently introduced breaking changes by adding a new field `connection_id` in the existing `Received` struct of their identify crate. Since the struct was modified and is used in `torii-relay`, it needs to be handled or ignored using `..` in the relevant match statement(s)

**Comment by kariy** [comment]
i think maybe bcs the your `libp2p` dependency got resolved to a different rev bcs in `torii-relay` the exact rev isn't specified https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/crates/torii/libp2p/Cargo.toml#L35-L43 .

so probably the best fix is to specify the exact rev of `libp2p` in `torii-relay` crate itself. based on `Cargo.lock` it's [451bcb6](https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/Cargo.lock#L8540)

**Comment by kariy** [comment]
@Larkooo any reason why we're using libp2p from git instead of crates.io ? 

https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/crates/torii/libp2p/Cargo.toml#L35-L37

**Comment by Larkooo** [comment]
> @Larkooo any reason why we're using libp2p from git instead of crates.io ? 
> 
> 
> 
> https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/crates/torii/libp2p/Cargo.toml#L35-L37

From what I recall I was having having some issues with official releases of libp2p that broke the trait attributes from some other deps that depend on it. That can easily be solved with a rev though so that should work just fine too

**Comment by elton-cs** [comment]
> > @Larkooo any reason why we're using libp2p from git instead of crates.io ?
> > https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/crates/torii/libp2p/Cargo.toml#L35-L37
> 
> From what I recall I was having having some issues with official releases of libp2p that broke the trait attributes from some other deps that depend on it. That can easily be solved with a rev though so that should work just fine too

Hmm so I'm a little confused about where exactly to add the rev. So assuming I add torii-relay as a dependency to a new library, you're saying I need to modify the Cargo.toml of the torii-relay package? as in modify torii/libp2p/Cargo.toml? 

If so, upon trying that, I got this error which seems to be exactly what you mentioned regarding the trait error:

<img width="1317" alt="image" src="https://github.com/dojoengine/dojo/assets/123023258/7ca28caa-110d-4ad6-8cf8-7fd05bc799bf">

**Comment by Larkooo** [comment]
> > > @Larkooo any reason why we're using libp2p from git instead of crates.io ?
> > > https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/crates/torii/libp2p/Cargo.toml#L35-L37
> > 
> > 
> > From what I recall I was having having some issues with official releases of libp2p that broke the trait attributes from some other deps that depend on it. That can easily be solved with a rev though so that should work just fine too
> 
> Hmm so I'm a little confused about where exactly to add the rev. So assuming I add torii-relay as a dependency to a new library, you're saying I need to modify the Cargo.toml of the torii-relay package? as in modify torii/libp2p/Cargo.toml?
> 
> If so, upon trying that, I got this error which seems to be exactly what you mentioned regarding the trait error:
> 
> <img alt="image" width="1317" src="https://private-user-images.githubusercontent.com/123023258/342747705-7ca28caa-110d-4ad6-8cf8-7fd05bc799bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMjQxMDUsIm5iZiI6MTcxOTMyMzgwNSwicGF0aCI6Ii8xMjMwMjMyNTgvMzQyNzQ3NzA1LTdjYTI4Y2FhLTExMGQtNGFkNi04Y2Y4LTdmZDA1YmM3OTliZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyNVQxMzU2NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNzUyMDUxMmZiYzRjMDlmODViZjk2YjgwNDBjNWY4YTVlNjU3YTY1ODg5ZDkwY2Q3OWIxZTMwYzI2N2ViYzYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.NQ8INJ6y5bNxpLxUHQTHP8cUkwO9G6dGr9PivKvas7U">
> Am I missing something here? Or maybe I misunderstood. Either way, some guidance would be very much appreciated 🙏

I just made a PR that should specify the rev for the libp2p and libp2p-webrtc dependencies. In your case, I believe, if you're also importing rust-libp2p or anything related to it that uses the same repo, you will need to specify the same rev. I remember it was super annoying to deal with the trait bounds as everything had to exactly match. I think the most we can do in torii-relay is just specify the revs

**Comment by glihm** [comment]
Closed by #2104.

**Comment by elton-cs** [comment]
> > > > @Larkooo any reason why we're using libp2p from git instead of crates.io ?
> > > > https://github.com/dojoengine/dojo/blob/8a5902b0074537f96384ea0829d54543adeb2367/crates/torii/libp2p/Cargo.toml#L35-L37
> > > 
> > > 
> > > From what I recall I was having having some issues with official releases of libp2p that broke the trait attributes from some other deps that depend on it. That can easily be solved with a rev though so that should work just fine too
> > 
> > 
> > Hmm so I'm a little confused about where exactly to add the rev. So assuming I add torii-relay as a dependency to a new library, you're saying I need to modify the Cargo.toml of the torii-relay package? as in modify torii/libp2p/Cargo.toml?
> > If so, upon trying that, I got this error which seems to be exactly what you mentioned regarding the trait error:
> > <img alt="image" width="1317" src="https://private-user-images.githubusercontent.com/123023258/342747705-7ca28caa-110d-4ad6-8cf8-7fd05bc799bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzMjQxMDUsIm5iZiI6MTcxOTMyMzgwNSwicGF0aCI6Ii8xMjMwMjMyNTgvMzQyNzQ3NzA1LTdjYTI4Y2FhLTExMGQtNGFkNi04Y2Y4LTdmZDA1YmM3OTliZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyNVQxMzU2NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNzUyMDUxMmZiYzRjMDlmODViZjk2YjgwNDBjNWY4YTVlNjU3YTY1ODg5ZDkwY2Q3OWIxZTMwYzI2N2ViYzYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.NQ8INJ6y5bNxpLxUHQTHP8cUkwO9G6dGr9PivKvas7U">
> > Am I missing something here? Or maybe I misunderstood. Either way, some guidance would be very much appreciated 🙏
> 
> I just made a PR that should specify the rev for the libp2p and libp2p-webrtc dependencies. In your case, I believe, if you're also importing rust-libp2p or anything related to it that uses the same repo, you will need to specify the same rev. I remember it was super annoying to deal with the trait bounds as everything had to exactly match. I think the most we can do in torii-relay is just specify the revs

Ahh that's what I was missing. I had updated just the libp2p dependency and not the related libp2p-webrtc ones! Thanks for the fix, appreciate it!
